net/tstun/mtu.go

```diff
@@ -1,33 +1,171 @@
 // Copyright (c) Tailscale Inc & AUTHORS
 // SPDX-License-Identifier: BSD-3-Clause
+
 package tstun
 
-import "tailscale.com/envknob"
+import (
+	"tailscale.com/envknob"
+)
+
+// The MTU (Maximum Transmission Unit) of a network interface is the largest
+// packet that can be sent or received through that interface, including all
+// headers above the link layer (e.g. IP headers, UDP headers, Wireguard
+// headers, etc.). We have to think about several different values of MTU:
+//
+// Wire MTU: The MTU of an interface underneath the tailscale TUN, e.g. an
```

Comment:
I feel like we still need to discuss the possibility of multiple wire paths to a peer somewhere.

Reply:
The MTU to a peer is stored per endpoint, so if I understand your comment correctly, that should allow us to have multiple different MTUs for different wire paths to a peer as long as they have different endpoints. We can still have multiple wire paths to the same IP/port but the standard TCP PMTUD implementation has to deal with that too, and I think we will handle it the same way: reprobe when we have too much packet loss and/or every 10 minutes. I think you might be suggesting that we could have multiple paths such that, e.g., every other packet goes over a path with one MTU and the in between packets go over a different path with another MTU? Which would definitely be bad for this algorithm. Let me know if I'm not getting it!

Reply:
Endpoint is an overloaded term in our system, but magicsock has an `endpoint` type that in our system is 1-1 with a peer, and internally carries IP:port tuples. When WireGuard sends a packet, it is directed to `(*endpoint).send` which then picks an address based on a latency sample and time based heuristic, falling back to DERP if there is no other good option.

We've observed that the selected IP:port in this selection can change regularly if there are influences to latency (for example, running a performance test that saturates bloated buffers along the link can trigger a switch in IP:port tuple regularly during the run). Most often for dual homed users this is a switch between IPv4 and IPv6, which would in fact alter the peer path MTU by the difference in header size between IPv4 and IPv6. This switching does not* cause an MTU change today because our MTU is fixed at 1280.

*Technically it does actually produce an MTU change, as it alters the minimum MTU space we need on the host link in order to successfully communicate packets, which if you happen to be sitting between 1300 and 1360 bytes MTU can be a problem, but we see this approximately never in practice. There appears to be a bias of either low MTUs (very close to or exactly 1280, such as the current WSL default), or higher MTUs (1400+).

Reply:
Notes from our discussion: first, yes, endpoint is overloaded :) I should have said "path MTU is stored per address" and there are multiple of those per endpoint/peer. Currently the MTU calculations assume worst case overhead (IPv6) so flipping between v4/v6 shouldn't break anything. We think that the worst case is we end up with the lower path MTU for 10 minutes, which is a good reason to avoid switching between best addresses unless there's a really obvious improvement. We think it's likely that if there's a link with a significantly larger MTU that it is also the lowest latency/highest bandwidth/most reliable path. I'll address this more in later commits as it doesn't need to be resolved in this one.